libraries/Microsoft.Bot.Builder/BotState.cs

```diff
@@ -6,34 +6,104 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Bot.Builder
 {
+
+    /** 
+     * State information cached off the context object by a `BotState` instance.
+     */
+    internal class CachedBotState
+    {
+        internal CachedBotState(IDictionary<string, object> state = null)
+        {
+            this.State = state ?? new Dictionary<string, object>();
+            this.Hash = ComputeHash(this.State);
+        }
+
+        internal IDictionary<string, object> State { get; set; }
+
+        internal string Hash { get; set; }
+
+        internal bool IsChanged()
+        {
+            return (Hash != ComputeHash(this.State));
+        }
+
+        private string ComputeHash(object obj)
+        {
+            return JsonConvert.SerializeObject(obj);
+        }
+    }
+
     /// <summary>
-    /// Base class which manages details of automatic loading and saving of bot state.
+    /// Reads and writes state for your bot to storage.
     /// </summary>
+    /// <remarks>
+    /// The state object will be automatically cached on the context object for the lifetime of the turn
+    /// and will only be written to storage if it has been modified.
+    /// 
+    /// When a `BotState` instance is used as middleware its state object will be automatically read in
+    /// before your bots logic runs and then intelligently written back out upon completion of your bots
+    /// logic. Multiple instances can be read and written in parallel using the `BotStateSet` middleware.
+    /// 
+    /// ```JavaScript
+    /// const { BotState, MemoryStorage
+    ///} require('botbuilder');
+    /// 
+    /// const storage = new MemoryStorage();
+    /// const botState = new BotState(storage, (context) => 'botState');
+    /// adapter.use(botState);
+    /// 
+    /// server.post('/api/messages', (req, res) => {
+    ///  *adapter.processActivity(req, res, async (context) => {
+    ///  *       // Track up time
+    ///  *       const state = botState.get(context);
+    ///  *       if (!('startTime' in state)) { state.startTime = new Date().getTime() }
+    ///  *state.upTime = new Date().getTime() - state.stateTime;
+    ///  *
+    ///  *       // ... route activity ...
+    ///  *
+    ///  *    });
+    /// });
+    /// ```
+    /// </remarks>
     /// <typeparam name="TState">The type of the bot state object.</typeparam>
-    public class BotState<TState> : IMiddleware
-        where TState : class, new()
+    public class BotState : IMiddleware, IPropertyContainer
```

Comment:
SRP warning: implementing `IPropertyContainer` direction on the middleware implementation gives a single class way too many responsibilities.

Suggest breaking the `IPropertyContainer` implementation details out to its own class that can be used/tested in isolation and then have this class create/control an instance of that class, but its primary focus can be on being a piece of middleware.